core/src/test/java/io/grpc/internal/DelayedStreamTest.java

```diff
@@ -0,0 +1,258 @@
+
+  @Mock private ClientStreamListener listener;
+  @Mock private ClientTransport transport;
+  @Mock private ClientStream realStream;
+  @Captor private ArgumentCaptor<Status> statusCaptor = ArgumentCaptor.forClass(Status.class);
+  private SettableFuture<ClientTransport> transportFuture = SettableFuture.create();
+  private DelayedStream stream;
+  private Metadata headers = new Metadata();
+
+  private MethodDescriptor<Integer, Integer> method = MethodDescriptor.create(
+      MethodType.UNARY, "service/method", new IntegerMarshaller(), new IntegerMarshaller());
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+
+    stream = new DelayedStream(this, CallOptions.DEFAULT, transportFuture, headers, listener,
+        executor, method);
+  }
+
+  @Test
+  public void realStreamCreatedImmediately() {
+    when(transport.newStream(
+        isA(MethodDescriptor.class),
+        isA(Metadata.class),
+        isA(ClientStreamListener.class)))
+        .thenReturn(realStream);
+    ListenableFuture<ClientTransport> future = Futures.immediateFuture(transport);
+
+    stream =
+        new DelayedStream(this, CallOptions.DEFAULT, future, headers, listener, executor, method);
+
+    verify(transport).newStream(method, headers, listener);
+  }
+
+  @Test
+  public void realStreamCreatedWhenTransportReady() {
+    when(transport.newStream(
+        isA(MethodDescriptor.class),
+        isA(Metadata.class),
+        isA(ClientStreamListener.class)))
+        .thenReturn(realStream);
+
+    // triggers createStream
+    transportFuture.set(transport);
+
+    verify(transport).newStream(method, headers, listener);
+  }
+
+  @Test
+  public void streamClosedOnTransportFailure() {
+    transportFuture.setException(new StatusRuntimeException(Status.INTERNAL));
+
+    verify(listener).closed(statusCaptor.capture(), isA(Metadata.class));
+    assertEquals(Status.Code.INTERNAL, statusCaptor.getValue().getCode());
+  }
+
+  @Test
+  public void streamClosedOnNullShutdownTransport() {
+    transportFuture.set(null);
+
+    verify(listener).closed(statusCaptor.capture(), isA(Metadata.class));
+    assertEquals(Status.Code.UNAVAILABLE, statusCaptor.getValue().getCode());
+    assertEquals("Channel is shutdown", statusCaptor.getValue().getDescription());
+  }
+
+  @Test
+  public void createStream_sendsAllMessages() {
+    when(transport.newStream(
+        isA(MethodDescriptor.class),
+        isA(Metadata.class),
+        isA(ClientStreamListener.class)))
+        .thenReturn(realStream);
+    stream.setCompressor(Codec.Identity.NONE);
+
+    DecompressorRegistry registry = DecompressorRegistry.newEmptyInstance();
+    stream.setDecompressionRegistry(registry);
+
+    stream.setMessageCompression(true);
+    InputStream message = new ByteArrayInputStream(new byte[]{'a'});
+    stream.writeMessage(message);
+    stream.setMessageCompression(false);
+    stream.writeMessage(message);
+
+    stream.createStream(transport);
```

Comment:
Calling `createStream()` from the test is too intrusive because it is meant to be called from within the class. If we make the change as @ejona86 has suggested, `setStream()` will be the interface that both the production and the test code would interact with, which feels better.